File: backend/data/ticketing.py
from datetime import date, datetime
from enum import Enum
from typing import ClassVar
from sqlmodel import Field, SQLModel

from backend.data import generate_uuid


class FareType(str, Enum):
    """Types of fares that can be applied to tickets."""

    STANDARD = "Standard"
    STUDENT = "Student"
    SENIOR = "Senior"
    DISCOUNT = "Discount"
    SURCHARGE = "Surcharge"
    OTHER = "Other"


class TicketCategory(str, Enum):
    """Categories of tickets based on their usage and validity."""

    TICKET = "Ticket"
    PASS = "Pass"
    OTHER = "Other"


class TicketType(SQLModel, table=True):
    __tablename__: ClassVar[str] = "ticket_types"

    id: str = Field(default_factory=generate_uuid, primary_key=True, index=True, max_length=36)
    name: str = Field(max_length=64, description="Name of the ticket type (e.g., 'Single Ride', 'Day Pass')")
    category: TicketCategory = Field(default=TicketCategory.TICKET, max_length=32)
    unit_price: int = Field(description="Base price of the ticket type in cents (e.g., 250 for $2.50)")
    valid_for_routes: str | None = Field(default=None, description="Comma-separated list of route IDs that the ticket is valid for (empty means all routes)")
    valid_for_rides: int | None = Field(default=None, description="Number of rides the ticket is valid for (null means unlimited)")
    valid_for_minutes: int | None = Field(default=None, description="Number of minutes the ticket is valid for after first validation (null means unlimited)")
    fare_type: FareType = Field(max_length=32)
    published: bool = Field(default=False, description="Indicates whether the ticket type is published and visible to customers for purchase (this locks the ticket type to further modifications)")
    active: bool = Field(default=False, description="Indicates whether the ticket type is currently active and available for purchase")

    description: str | None = Field(default=None, description="Public description of the ticket type")
    private_notes: str | None = Field(default=None, description="Internal notes not visible to customers")


class IssuedTicket(SQLModel, table=True):
    __tablename__: ClassVar[str] = "tickets_issued"

    id: str = Field(default_factory=generate_uuid, primary_key=True, index=True, max_length=36)
    ticket_type_id: str = Field(foreign_key="ticket_types.id", index=True, description="Link to the ticket type")
    customer_id: str = Field(foreign_key="customers.id", index=True, description="Link to the customer who owns the ticket")
    issued_at: datetime = Field(max_length=26)
    expires_at: datetime = Field(max_length=26)

    fare_type: FareType = Field(max_length=32)
    final_price: int = Field(description="Final price paid by the customer for the ticket in cents (e.g., 250 for $2.50)")
    validated_at: datetime = Field(max_length=26, description="Timestamp of the first validation of the ticket (null if never validated)")
    flagged: bool = Field(default=False, description="Indicates whether the ticket has been flagged for review due to validation issues or suspected fraud")


class TicketValidation(SQLModel, table=True):
    __tablename__: ClassVar[str] = "ticket_validations"

    id: str = Field(default_factory=generate_uuid, primary_key=True, index=True, max_length=36)
    issued_ticket_id: str = Field(foreign_key="tickets_issued.id", index=True, description="Link to the issued ticket that was validated")
    device_id: str = Field(foreign_key="devices.id", index=True, description="Link to the device where the ticket was validated")
    vehicle_id: str = Field(foreign_key="vehicles.id", index=True, description="Link to the vehicle where the ticket was validated")
    validated_at: datetime = Field(max_length=26, description="Timestamp of the validation of the ticket")
    validation_result: str = Field(max_length=255, description="Result of the validation (e.g., 'Valid', 'Expired', 'Invalid')")

File: mobile/types/ticketing.ts
export enum FareType {
    STANDARD = "Standard",
    STUDENT = "Student",
    SENIOR = "Senior",
    DISCOUNT = "Discount",
    SURCHARGE = "Surcharge",
    OTHER = "Other",
}

export enum TicketCategory {
    TICKET = "Ticket",
    PASS = "Pass",
    OTHER = "Other",
}

export interface TicketType {
    id: string;
    name: string;
    category: TicketCategory;
    unit_price: number; // Price in cents
    valid_for_routes?: string; // Comma-separated route IDs
    valid_for_rides?: number;
    valid_for_minutes?: number;
    fare_type: FareType;
    published: boolean;
    active: boolean;
    description?: string;
    private_notes?: string;
}

export interface IssuedTicket {
    id: string;
    ticket_type_id: string;
    customer_id: string;
    issued_at: string; // ISO 8601 format
    expires_at: string; // ISO 8601 format
    fare_type: FareType;
    final_price: number; // Price in cents
    validated_at?: string; // ISO 8601 format
    flagged: boolean;
}

export interface TicketValidation {
    id: string;
    issued_ticket_id: string;
    device_id: string;
    vehicle_id: string;
    validated_at: string; // ISO 8601 format
    validation_result: string;
}

export interface CustomerWallet {
    id: string;
    customer_id: string;
    balance: number; // Balance in cents
    last_topup_at?: string; // ISO 8601 format
    active: boolean;
}

export enum WalletTransactionType {
    TOPUP = "Top-up",
    PURCHASE = "Purchase",
    REFUND = "Refund",
    OTHER = "Other",
}

export interface CustomerWalletTransaction {
    id: string;
    wallet_id: string;
    amount: number; // Amount in cents
    transaction_type: WalletTransactionType;
    created_at: string; // ISO 8601 format
    flagged: boolean;
}
